{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Colocalization App Tutorial\n",
    "\n",
    "This tutorial will give you an idea of the backing algorithm to our Colocalization app by using a toy dataset.\n",
    "\n",
    "This project allows researchers to colocalize eQTL hits from eQTLGen and GWAS hits from any study in openGWAS. It operates under the assumption that some GWAS and eQTL hits in close proximity may be the same signal, linking phenotypes to gene expression and enabling the development of transcriptional risk scores (TRS) and polygenic-predicted transcriptional risk scores (PP-TRS) for these phenotypes. As input, the project takes any study ID from <a href=\"https://www.opengwas.org/\">openGWAS</a> as well as eQTL summary statistics from <a href=\"https://www.eqtlgen.org/cis-eqtls.html\">eQTLGen</a>. The project has been implemented as an <a href=\"https://genapp2022.biosci.gatech.edu/team1/\">app</a> using whole blood eQTL summary statistics, allowing users to download a list of GWAS and eQTL hits found to be the same signal. It was originally developed as a class project for BIOL8803 at Georgia Tech in Fall 2022.\n",
    "\n",
    "Use this link to access the public Github which contains the necessary files to run the procedure locally and a detailed README: https://github.com/knishiura3/BIOL8803_Transcriptional_Risk_Score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup\n",
    "Ensure that you've set R as your kernel in Jupyter. See this tutorial for how to obtain an R kernel:\n",
    "https://towardsdatascience.com/how-to-run-r-scripts-in-jupyter-15527148d2a\n",
    "\n",
    "Run the following code blocks when the after completing the steps outlined in the README file on the command line. This ensures your conda environment is suitable for running the app. Verify that you have both the LD and MAF files locally as well.\n",
    "\n",
    "This will also load the necessary packages from that environment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# packages for parquet querying\n",
    "library(arrow)\n",
    "library(duckdb)\n",
    "library(fs)\n",
    "library(tidyverse)\n",
    "library(DBI)\n",
    "library(glue)\n",
    "\n",
    "# packages for gassocplot\n",
    "library(grid)\n",
    "library(gridExtra) \n",
    "library(gtable)\n",
    "\n",
    "\n",
    "# packages for coloc\n",
    "suppressPackageStartupMessages(suppressWarnings({\n",
    "    library(gwasglue)\n",
    "    library(gassocplot)\n",
    "    library(dplyr)\n",
    "    library(coloc)\n",
    "    library(ggplot2)\n",
    "    library(httpgd)\n",
    "\n",
    "}))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define helper functions\n",
    "Within the app, these are sourced from three separate R files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#from eQTL_query_parquet.r\n",
    "initialize_db <- function() {\n",
    "    # open parquet files\n",
    "    ds_eQTL <- arrow::open_dataset(dir_eqtl, partitioning = \"SNPChr\")\n",
    "    ds_eqtlMAF <- arrow::open_dataset(dir_eqtlmaf, partitioning = \"hg19_chr\")\n",
    "\n",
    "    # clean up if previous connection if still open (only happens if script is interrupted)\n",
    "    if (exists(\"con\")) {\n",
    "        duckdb::duckdb_unregister(con, \"eqtlTable\")\n",
    "        duckdb::duckdb_unregister(con, \"mafTable\")\n",
    "        duckdb::dbDisconnect(con)\n",
    "    }\n",
    "\n",
    "    # connect duckdb to open parquet files\n",
    "    con <- duckdb::dbConnect(duckdb::duckdb())\n",
    "\n",
    "    # register the datasets as DuckDB table, and give it a name\n",
    "    duckdb::duckdb_register_arrow(con, \"eqtlTable\", ds_eQTL)\n",
    "    duckdb::duckdb_register_arrow(con, \"mafTable\", ds_eqtlMAF)\n",
    "    return(con)\n",
    "}\n",
    "\n",
    "gather_and_format_gwas_eqtl_in_region <- function(con, chromosome, gwas_hit) {\n",
    "    # str(gwas_hit)\n",
    "    # get the position of the current GWAS hit\n",
    "    gwas_pos <- gwas_hit$position\n",
    "    lower <- gwas_pos - window_size\n",
    "    # set floor of 0 for lower bound of window\n",
    "    if (lower < 0) {\n",
    "        lower <- 0\n",
    "    }\n",
    "    upper <- gwas_pos + window_size\n",
    "    chrpos <- paste0(chromosome, \":\", lower, \"-\", upper)\n",
    "\n",
    "    # window function query for top eQTLs, defined as:\n",
    "    #   for each locus, determined by values of chromosome & position,\n",
    "    #   sort descending by absolute value Zcore and take only top row)\n",
    "    subquery_eqtl <- glue::glue(\n",
    "        \"\n",
    "        SELECT * FROM(\n",
    "            SELECT *, ROW_NUMBER() OVER (PARTITION BY SNPPos ORDER BY abs(Zscore) DESC) AS 'row'\n",
    "            FROM eqtlTable\n",
    "            WHERE SNPChr = {chromosome} AND SNPPos BETWEEN {lower} AND {upper}\n",
    "            )\n",
    "        WHERE row = 1\n",
    "        \"\n",
    "    )\n",
    "    subquery_maf <- glue::glue(\n",
    "        \"\n",
    "        SELECT * FROM mafTable\n",
    "        WHERE hg19_chr = {chromosome}\n",
    "        \"\n",
    "    )\n",
    "    # execute the SQL queries\n",
    "    result_eqtl <- dbGetQuery(con, subquery_eqtl)\n",
    "    result_maf <- dbGetQuery(con, subquery_maf)\n",
    "    print('SQL queries executed')\n",
    "    # left join the eQTL and MAF tables to get full set of relevant eqtl data\n",
    "    result_raw <- left_join(result_eqtl, result_maf,\n",
    "        by = c(\"SNPChr\" = \"hg19_chr\", \"SNPPos\" = \"hg19_pos\"),\n",
    "        suffix = c(\".eqtl\", \".maf\")\n",
    "    ) \n",
    "    print('join complete')\n",
    "    # filter and rename columns\n",
    "    result <- result_raw %>%\n",
    "        # keep columns Pvalue, SNP, SNPPos, Zscore, NrSamples, AlleleB_all\n",
    "        dplyr::select(Pvalue, NrSamples, AlleleB_all, SNP.eqtl, Zscore, SNPPos, AssessedAllele,OtherAllele) %>%\n",
    "        # add back column SNPChr\n",
    "        tibble::add_column(chr = as.character(chromosome), .before = \"SNPPos\") %>%\n",
    "        # rename AlleleB_all to MAF\n",
    "        dplyr::rename(eaf = AlleleB_all, rsid = SNP.eqtl, p = Pvalue, n = NrSamples, z = Zscore, position = SNPPos, ea = AssessedAllele, nea = OtherAllele)\n",
    "    \n",
    "    # str(result)\n",
    "    # harmonize gwas and eqtl datasets and structure for downstream analysis  \n",
    "    out <- ieugwasr_to_coloc_custom(\n",
    "        id1= gwas_dataset, \n",
    "        result, # eqtl data\n",
    "        type2 = \"quant\", \n",
    "        chrompos = as.character(chrpos)\n",
    "    )\n",
    "    \n",
    "    # str(out)\n",
    "    # run coloc\n",
    "    res <- coloc::coloc.abf(out[[1]], out[[2]])\n",
    "    # str(res)\n",
    "    # extract H4 from coloc output\n",
    "    PP_H4 <- round(as.numeric(res$summary[[\"PP.H4.abf\"]]), digits = 2)\n",
    "    \n",
    "    return(list(out, PP_H4, result_raw))\n",
    "}\n",
    "\n",
    "extract_top_markers_from_gassocplot_output <- function(temp) {\n",
    "    # str(temp)\n",
    "    # code modified from gassocplot for identifying top markers\n",
    "    # https://github.com/jrs95/gassocplot/blob/master/R/figures.R#L387-L390\n",
    "    mlog10p_gwas <- -(log(2) + pnorm(-abs(temp$z[[1]]), log.p=TRUE))/log(10)\n",
    "    mlog10p_eqtl <- -(log(2) + pnorm(-abs(temp$z[[2]]), log.p=TRUE))/log(10)\n",
    "    # add as column of temp object because of shared indices\n",
    "    temp$stats_gwas <- mlog10p_gwas\n",
    "    temp$stats_eqtl <- mlog10p_eqtl\n",
    "    # get the values and indices of the the maximum value of temp$stats (or elements of values/indices tied for max)\n",
    "    max_pval_gwas <- max(temp$stats_gwas)\n",
    "    max_pval_eqtl <- max(temp$stats_eqtl)\n",
    "    # str(max_pval_eqtl)\n",
    "    max_pval_indices_gwas <- which(temp$stats_gwas == max_pval_gwas)\n",
    "    max_pval_indices_eqtl <- which(temp$stats_eqtl == max_pval_eqtl)\n",
    "    # str(max_pval_indices_eqtl)\n",
    "    # get the rsid (temp$markers$marker) at the index position of top marker\n",
    "    top_marker_gwas <- temp$markers$marker[max_pval_indices_gwas]\n",
    "    top_marker_eqtl <- temp$markers$marker[max_pval_indices_eqtl]\n",
    "\n",
    "    return(list(top_marker_gwas, top_marker_eqtl))\n",
    "}\n",
    "\n",
    "# take top markers as input and return a raw data row for each colocalized gwas and eqtl pair\n",
    "get_top_marker_raw_data <- function(top_marker_gwas, top_marker_eqtl, result_raw, PP_H4) {\n",
    "    # create a table of the row in result_raw that corresponds to the top marker and append H4 as a column\n",
    "    # and only keep desired columns\n",
    "    top_eqtl_table <- result_raw[result_raw$SNP.eqtl %in% top_marker_eqtl, c(\"GeneChr\",\"Pvalue\", \"SNP.eqtl\", \"SNPPos\", \"AssessedAllele\", \"OtherAllele\", \"Zscore\", \"Gene\", \"GenePos\", \"NrCohorts\", \"NrSamples\", \"FDR\", \"BonferroniP\", \"AlleleA\", \"AlleleB\", \"allA_total\", \"allAB_total\", \"allB_total\", \"AlleleB_all\")]\n",
    "    # rename GeneChr to Chr\n",
    "    colnames(top_eqtl_table)[colnames(top_eqtl_table) == \"GeneChr\"] <- \"Chr\"\n",
    "    # loop over and add .eqtl suffix to columns Pvalue, SNPPos, AssessedAllele, OtherAllele, Zscore, AlleleA, AlleleB, allA_total, allAB_total, allB_total, AlleleB_all\n",
    "    for (i in 1:ncol(top_eqtl_table)) {\n",
    "        # skip column name that already has .eqtl suffix or if Chr\n",
    "        if (grepl(\".eqtl\", colnames(top_eqtl_table)[i]) | colnames(top_eqtl_table)[i] == \"Chr\") {\n",
    "            next\n",
    "        }\n",
    "        colnames(top_eqtl_table)[i] <- paste0(colnames(top_eqtl_table)[i], \".eqtl\")\n",
    "    }\n",
    "\n",
    "    # queries GWAS table w/ rsid for top marker\n",
    "    top_gwas <- ieugwasr::associations(top_marker_gwas,gwas_dataset)\n",
    "    # only keep columns position, beta, se, p, n, id, rsid, ea, nea, eaf, trait\n",
    "    top_gwas <- top_gwas[, c(\"position\", \"beta\", \"se\", \"p\", \"n\", \"id\", \"rsid\", \"ea\", \"nea\", \"eaf\", \"trait\")]\n",
    "    colnames(top_gwas) <- paste0(colnames(top_gwas), \".gwas\")\n",
    "    # combine top_gwas and top_eqtl_table side-by-side in new table\n",
    "    top_table <- cbind(top_gwas, top_eqtl_table)\n",
    "    # move Chr column to first position\n",
    "    top_table <- top_table[, c(\"Chr\", colnames(top_table)[!colnames(top_table) %in% \"Chr\"])]\n",
    "    top_table$H4 <- PP_H4\n",
    "    return(top_table)\n",
    "}\n",
    "\n",
    "plot_associated_signals_and_save <- function(tophit_idx, chromosome, PP_H4, temp, plot_dir) {\n",
    "    # this is needed to suppress automatic pdf output from gassocplot\n",
    "    pdf(NULL)\n",
    "    # construct plot\n",
    "    theplot <- stack_assoc_plot_custom(temp$markers, temp$z, temp$corr, traits = temp$traits)\n",
    "    # output path for saving a figure for each colocalized pair of GWAS/eQTL \n",
    "    outfile <- glue::glue(plot_dir, \"/chr{chromosome}_gwastophit{sprintf('%03d', tophit_idx)}_H4_{PP_H4}.png\")\n",
    "    # if it doesn't exist, create directory\n",
    "    if (!dir.exists(plot_dir)) {\n",
    "        dir.create(plot_dir, recursive = TRUE)\n",
    "    }\n",
    "    # save plot w/ base R functions\n",
    "    \n",
    "    png(filename = outfile)\n",
    "    plot(theplot)\n",
    "    dev.off()\n",
    "}\n",
    "\n",
    "# function to clean up\n",
    "clean_up <- function(con) {\n",
    "    # clean up\n",
    "    print(\"cleaning up\")\n",
    "    duckdb_unregister(con, \"eqtlTable\")\n",
    "    duckdb_unregister(con, \"mafTable\")\n",
    "    dbDisconnect(con, shutdown = TRUE)\n",
    "    print(\"done\")\n",
    "}\n",
    "\n",
    "\n",
    "#from ieugwasr_to_coloc.r\n",
    "ieugwasr_to_coloc_custom <- function(id1, eqtl_df, chrompos, type1=NULL, type2=NULL)\n",
    "{\n",
    "    id2 <- 'eQTL-Gen'\n",
    "\ttab1 <- ieugwasr::associations(id=id1, variants=chrompos) %>% subset(., !duplicated(rsid))\n",
    "\t# tab2 <- ieugwasr::associations(id=id2, variants=chrompos) %>% subset(., !duplicated(rsid))\n",
    "\ttab2 <- eqtl_df %>% subset(., !duplicated(rsid)) # tab2 is the eQTL dataset\n",
    "\tcommonsnps <- tab1$rsid[tab1$rsid %in% tab2$rsid]\n",
    "\ttab1 <- tab1[tab1$rsid %in% commonsnps, ] %>% dplyr::arrange(rsid)\n",
    "\ttab2 <- tab2[tab2$rsid %in% commonsnps, ] %>% dplyr::arrange(rsid)\n",
    "\tstopifnot(all(tab1$rsid == tab2$rsid))\n",
    "\n",
    "\tindex <- as.character(tab1$ea) == as.character(tab2$ea) &\n",
    "\t\t\tas.character(tab1$nea) == as.character(tab2$nea) &\n",
    "\t\t\tas.character(tab1$rsid) == as.character(tab2$rsid) &\n",
    "\t\t\ttab1$position == tab2$position\n",
    "\tstopifnot(sum(index) > 0)\n",
    "\ttab1$eaf <- as.numeric(tab1$eaf)\n",
    "\ttab2$eaf <- as.numeric(tab2$eaf)\n",
    "\ttab1$eaf[which(tab1$eaf > 0.5)] <- 1 - tab1$eaf[which(tab1$eaf > 0.5)]\n",
    "\ttab2$eaf[which(tab2$eaf > 0.5)] <- 1 - tab2$eaf[which(tab2$eaf > 0.5)]\n",
    "\ts <- sum(is.na(tab1$eaf))\n",
    "\tif(s > 0)\n",
    "\t{\n",
    "\t\twarning(s, \" out of \", nrow(tab1), \" variants have missing allele frequencies in \", id1, \". Setting to 0.5\")\n",
    "\t\ttab1$eaf[is.na(tab1$eaf)] <- 0.5\n",
    "\t}\n",
    "\ts <- sum(is.na(tab2$eaf))\n",
    "\tif(s > 0)\n",
    "\t{\n",
    "\t\twarning(s, \" out of \", nrow(tab2), \" variants have missing allele frequencies in \", id2, \". Setting to 0.5\")\n",
    "\t\ttab2$eaf[is.na(tab2$eaf)] <- 0.5\n",
    "\t}\n",
    "\n",
    "\tinfo1 <- ieugwasr::gwasinfo(id1)\n",
    "\ttype1 <- get_type(info1, type1)\n",
    "\t# info2 <- ieugwasr::gwasinfo(id2)\n",
    "\ttype2 <- \"quant\"\n",
    "\n",
    "\n",
    "\ttab1$n[is.na(tab1$n)] <- info1$sample_size\n",
    "    # swap any NA values in eqtl table w/ median sample size\n",
    "    tab2$n[is.na(tab2$n)] <- median(tab2$n)\n",
    "\n",
    "\ttab1 <- tab1[index,] %>% {list(pvalues = .$p, N = .$n, MAF = .$eaf, beta = .$beta, varbeta = .$se^2, type = type1, snp = .$rsid, z = .$beta / .$se, chr = .$chr, pos = .$position, id = id1)}\n",
    "\ttab2 <- tab2[index,] %>% {list(pvalues = .$p, N = .$n, MAF = .$eaf, \n",
    "        # beta = NULL, varbeta = NULL, \n",
    "        type = type2, snp = .$rsid, z = .$z, chr = .$chr, pos = .$position, id = id2)}\n",
    "\n",
    "\tif(type1 == \"cc\")\n",
    "\t{\n",
    "\t\ttab1$s <- info1$ncase / info1$sample_size\n",
    "\t}\n",
    "\n",
    "\tif(type2 == \"cc\")\n",
    "\t{\n",
    "\t\ttab2$s <- info2$ncase / info2$sample_size\n",
    "\t}\n",
    "\n",
    "\treturn(list(dataset1=tab1, dataset2=tab2))\n",
    "}\n",
    "\n",
    "get_type <- function(info, typex)\n",
    "{\n",
    "\tif(!is.null(typex))\n",
    "\t{\n",
    "\t\tstopifnot(typex %in% c(\"cc\", \"quant\"))\n",
    "\t\treturn(typex)\n",
    "\t} else if(is.na(info$unit)) {\n",
    "\t\tif(! \"ncase\" %in% names(info))\n",
    "\t\t{\n",
    "\t\t\tinfo$ncase <- NA\n",
    "\t\t}\n",
    "\t\tif(is.na(info$ncase))\n",
    "\t\t{\n",
    "\t\t\tmessage(\"Type information not available for \", info$id, \". Assuming 'quant' but override using 'type' arguments.\")\n",
    "\t\t\treturn(\"quant\")\t\t\t\n",
    "\t\t} else {\n",
    "\t\t\tmessage(\"No units available but assuming cc due to number of cases being stated\")\n",
    "\t\t\treturn(\"cc\")\n",
    "\t\t}\n",
    "\t} else {\n",
    "\t\treturn(ifelse(info$unit %in% c(\"logOR\", \"log odds\"), \"cc\", \"quant\"))\n",
    "\t}\n",
    "}\n",
    "\n",
    "\n",
    "#from gassocplot.r\n",
    "g_legend<-function(gplot){\n",
    "  tmp <- ggplot_gtable(ggplot_build(gplot))\n",
    "  leg <- which(sapply(tmp$grobs, function(x) x$name) == \"guide-box\")\n",
    "  legend <- tmp$grobs[[leg]]\n",
    "  return(legend)\n",
    "}\n",
    "\n",
    "add_g_legend <- function(g, legend){\n",
    "  lheight <- sum(legend$height)*1.5\n",
    "  g <- grid.arrange(g, legend, ncol = 1, heights = unit.c(unit(1, \"npc\") - lheight, lheight))\n",
    "  return(g)\n",
    "}\n",
    "\n",
    "stack_assoc_plot_custom <- function(markers, z, corr=NULL, corr.top=NULL, traits, ylab=NULL, type=\"log10p\", x.min=NULL, x.max=NULL, top.marker=NULL, legend=TRUE){\n",
    "  \n",
    "  # Error messages\n",
    "  if(!(type==\"log10p\" | type==\"prob\")) stop(\"the type of plot has to be either log10p or prob\")\n",
    "  if(length(traits)!=ncol(z)) stop(\"the number of traits is not the same as the number of columns for the Z-scores\")\n",
    "  if(nrow(markers)!=nrow(z)) stop(\"the number of markers is not the same as the number of rows for the Z-scores\")\n",
    "  if(!is.null(corr)){if(ncol(corr)!=nrow(markers) | nrow(corr)!=nrow(markers)) stop(\"corr has to have the same dimensions as the number of rows in the markers dataset\")}\n",
    "  if(!is.null(corr.top)){if(length(corr.top)!=nrow(markers)) stop(\"corr.top has to have the same length as the number of rows in the markers dataset\")}\n",
    "  # if(any(rownames(corr)!=markers$marker)) stop(\"corr has to have the same markers in the same order as the markers dataset\")\n",
    "  if(any(names(markers)!=c(\"marker\", \"chr\", \"pos\"))) stop(\"dataset needs to include marker, chr and pos columns in that order\")\n",
    "  if(length(unique(markers$chr))>1) stop(\"there should only be markers from one chromosome in the markers dataset\")   \n",
    "  if(!(markers$chr[1] %in% 1:22)) stop(\"the plotting tool is only for autosomal chromosomes\") \n",
    "  if(any(is.na(markers))) stop(\"there are missing markers in your marker dataset\") \n",
    "  # if(any(is.na(z))) stop(\"there are missing values in the Z-score matrix\")\n",
    "  if(class(markers$pos)!=\"integer\") stop(\"the pos variable has to be an integer\")\n",
    "  if(is.null(corr) & !is.null(corr.top) & is.null(top.marker)) stop(\"top.marker must be defined if corr.top is provided\")\n",
    "  if(is.null(corr) & !is.null(corr.top)){if(length(corr.top)!=nrow(markers)) stop(\"corr.top has to have the same length as the number of rows in the markers dataset\")}\n",
    "  if(!is.null(top.marker) & length(which(top.marker==markers$marker))==0) stop(\"top.marker is not contained in the markers dataset\")\n",
    "  if(!is.null(top.marker) & length(which(top.marker==markers$marker))>1) stop(\"top.marker maps to multiple markers in the markers dataset\")\n",
    "  \n",
    "  # Coerce data\n",
    "  markers$marker <- as.character(markers$marker)\n",
    "  chr <- as.integer(markers$chr[1])\n",
    "  r2_legend <- legend\n",
    "  if(is.null(x.min)){x.min <- min(as.integer(markers$pos))}\n",
    "  if(is.null(x.max)){x.max <- max(as.integer(markers$pos))}\n",
    "  if((x.max - x.min)>10000000) stop(\"the plotting tool can plot a maximum of 10MB\")\n",
    "  \n",
    "  # mlog10p\n",
    "  if(type==\"log10p\"){\n",
    "    mlog10p <- suppressWarnings(apply(z, 2, function(x){-(log(2) + pnorm(-abs(x), log.p=T))/log(10)}))\n",
    "    # Kenji Nishiura:  don't impose a -log10(pval) ceiling \n",
    "    # mlog10p[mlog10p>1000 & !is.na(mlog10p)] <- 1000\n",
    "  }\n",
    "  if(type==\"log10p\"){ylab <- expression(\"-log\"[\"10\"]*paste(\"(\",italic(\"p\"),\")\"))}else{if(is.null(ylab)){ylab <- \"Probability\"}}\n",
    " \n",
    "  # Genes\n",
    "  gene.region <- gassocplot::genes[gassocplot::genes$chr==chr & !(gassocplot::genes$end<x.min) & !(gassocplot::genes$start>x.max),]\n",
    "  gene.region$start[gene.region$start<x.min] <- x.min\n",
    "  gene.region$end[gene.region$end>x.max] <- x.max\n",
    "  gene.region <- gene.region[with(gene.region, order(start)), ]\n",
    "  ngenes <- nrow(gene.region)\n",
    "\n",
    "  # Max and min\n",
    "  x.min <- x.min - 0.02*(x.max - x.min)\n",
    "  x.max <- x.max + 0.02*(x.max - x.min)\n",
    "  \n",
    "  # Correlation matrix\n",
    "  if(is.null(corr) & is.null(corr.top)){r2_legend <- FALSE; corr <- matrix(NA, nrow=nrow(markers), ncol=nrow(markers))}\n",
    "\n",
    "  # Recombination plot\n",
    "  recombination.plot <- plot_recombination_rate_stack(chr, x.min, x.max)\n",
    "\n",
    "  # Gene plot\n",
    "  if(ngenes==0){gene.plot <- plot_gene_zero(chr, x.min, x.max, stack=TRUE)}\n",
    "  if(ngenes>0 & ngenes<=5){gene.plot <- plot_gene_two(gene.region, chr, x.min, x.max, stack=TRUE)}\n",
    "  if(ngenes>5 & ngenes<=10){gene.plot <- plot_gene_five(gene.region, chr, x.min, x.max, stack=TRUE)}\n",
    "  if(ngenes>10 & ngenes<=25){gene.plot <- plot_gene_ten(gene.region, chr, x.min, x.max, stack=TRUE)}\n",
    "  if(ngenes>25){gene.plot <- plot_gene_fifteen(gene.region, chr, x.min, x.max, stack=TRUE)}\n",
    "\n",
    "  # Top marker\n",
    "  if(length(top.marker)!=0){if(is.na(top.marker)){top.marker <- NULL}}\n",
    "    \n",
    "  # Association plot\n",
    "  for(i in length(traits):1){\n",
    "    if(type==\"log10p\"){\n",
    "      data <- data.frame(marker=markers$marker, chr=as.integer(markers$chr), pos=as.integer(markers$pos), stats=mlog10p[,i], stringsAsFactors=F)\n",
    "    }else{\n",
    "      data <- data.frame(marker=markers$marker, chr=as.integer(markers$chr), pos=as.integer(markers$pos), stats=z[,i], stringsAsFactors=F)    \n",
    "    }\n",
    "    marker.plot <- plot_assoc_stack(data, corr, corr.top, x.min, x.max, top.marker, ylab, type)\n",
    "    legend <- g_legend(marker.plot)\n",
    "    if(i==length(traits)){g <- plot_regional_gene_assoc(recombination.plot, marker.plot, gene.plot, traits[i], ngenes)}\n",
    "    if(i<length(traits)){\n",
    "      g1 <- plot_regional_assoc(recombination.plot, marker.plot, traits[i])\n",
    "      g <- gtable:::rbind_gtable(g1, g, \"last\")\n",
    "      panels <- g$layout$t[grep(\"panel\", g$layout$name)]\n",
    "      g$heights[panels[1]] <- unit(3,\"null\") \n",
    "    } \n",
    "  }\n",
    "\n",
    "  # Combined plot\n",
    "  if(r2_legend==T){\n",
    "    combined.plot <- add_g_legend(g, legend)\n",
    "  }else{\n",
    "    combined.plot <- g\n",
    "  }\n",
    "\n",
    "  return(combined.plot)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collect Inputs\n",
    "Within the app, this is done reactively in R Shiny whenever an interactive field is changed and the submit button is pressed. In this tutorial, we'll use ieu-b-30 as the example GWAS ID and focus on chromosome 21."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_dataset <<- \"ieu-b-30\"\n",
    "chromosomes <<- 21\n",
    "dir_ld <<- \"ld\"\n",
    "dir_eqtl <<- \"data/eqtls\"\n",
    "dir_eqtlmaf <<- \"data/eqtl_MAF\"\n",
    "eqtl_outdir <<- \"top_eqtls\"\n",
    "plot_dir <<- \"plots\"\n",
    "\n",
    "# if output subdirectories don't exist yet, create them\n",
    "if (!dir.exists(eqtl_outdir)) {\n",
    "    dir.create(eqtl_outdir)\n",
    "}\n",
    "if (!dir.exists(plot_dir)) {\n",
    "    dir.create(plot_dir)\n",
    "}\n",
    "\n",
    "# if it doesn't exist, create a directory named user_outdir in plots and top_eqtls\n",
    "# if it exists already, delete files from previous run\n",
    "# if user_outdir is defined and the plot_dir/user_outdir exists, delete the files in that directory\n",
    "if (exists(\"user_outdir\") && dir.exists(glue::glue(plot_dir,'/', user_outdir))) {\n",
    "    unlink(glue::glue(plot_dir,'/', user_outdir, '/*'))\n",
    "    unlink(glue::glue(eqtl_outdir,'/', user_outdir, '/*'))\n",
    "}\n",
    "\n",
    "# generate unique id using timestamp\n",
    "user_outdir <<- as.character(Sys.time())\n",
    "user_outdir <<- gsub(\" \", \"_\", user_outdir)\n",
    "user_outdir <<- gsub(\":\", \"_\", user_outdir)\n",
    "user_outdir <<- gsub(\"-\", \"_\", user_outdir)\n",
    "\n",
    "# if user subdirectories don't exist yet, create them\n",
    "if (!dir.exists(glue::glue(plot_dir,'/', user_outdir))) {\n",
    "    dir.create(glue::glue(plot_dir,'/', user_outdir))\n",
    "}\n",
    "if (!dir.exists(glue::glue(eqtl_outdir,'/', user_outdir))) {\n",
    "    dir.create(glue::glue(eqtl_outdir,'/', user_outdir))\n",
    "}\n",
    "\n",
    "\n",
    "window_size <<- as.integer(100000)\n",
    "con <- initialize_db()\n",
    "# query API for top GWAS hits for all chromosomes\n",
    "top_all <- ieugwasr::tophits(gwas_dataset)\n",
    "\n",
    "# initialize counter at 0 each time submit button is pressed\n",
    "colocalized_counter <- 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run colocalization analysis on each chromosome\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for (chromosome in chromosomes) {    \n",
    "    # filter the top GWAS hits for the current chromosome\n",
    "    top <- top_all %>%\n",
    "        filter(chr == chromosome) %>%\n",
    "        arrange(p)\n",
    "    # initialize progressbar for each chromosome separately\n",
    "#     withProgress(\n",
    "#         message='Please wait',\n",
    "#         detail=glue::glue('Processing chr {chromosome}:'),\n",
    "#         value=0, {\n",
    "    for (tophit_idx in seq_len(nrow(top))) {\n",
    "        # calls customized ieugwasr_to_coloc function to take in eQTL data as direct input\n",
    "        out_PPH4_rawResult <- gather_and_format_gwas_eqtl_in_region(con, chromosome, top[tophit_idx,])\n",
    "        # unpack returned objects\n",
    "        out <- out_PPH4_rawResult[[1]]\n",
    "        PP_H4 <- out_PPH4_rawResult[[2]]\n",
    "        # continue to the next window if posterior probability of H4 is less than 0.5\n",
    "        if (PP_H4 < 0.50) {\n",
    "            print(glue::glue(\"PP_H4 < 0.50, skipping and continuing to next GWAS top hit\"))\n",
    "            setProgress(tophit_idx / nrow(top), detail = glue::glue('Processing chr {chromosome}: region {tophit_idx} out of {nrow(top)}'))\n",
    "            next\n",
    "        }\n",
    "        # API rejects requests if >500 rsids, so need to run plink locally\n",
    "        if (length(out[[2]]$pos) >= 500) {\n",
    "            print(\"too many rsids, running plink locally\")\n",
    "            # input to coloc_to_gassocplot is list of rsids (should be identical in gwas/eqtl data at this stage): out[[1]]$snp\n",
    "            # choices for ancestry are AMR, AFR, EAS, EUR, SAS\n",
    "            # note: a bit slow first time because the plink_bin function will download/install plink if it's not already installed.\n",
    "            temp <- coloc_to_gassocplot(out, bfile = paste0(dir_ld, \"/EUR\"), plink_bin = '/projects/team1/bin/mambaforge/envs/plink/bin/plink')\n",
    "            # local debugging (comment out line above, uncomment line below)\n",
    "            # temp <- coloc_to_gassocplot(out, bfile = paste0(dir_ld, \"/EUR\"), plink_bin = genetics.binaRies::get_plink_binary())\n",
    "            # increment counter for number of colocalized eQTLs\n",
    "            colocalized_counter <- colocalized_counter + 1\n",
    "        } else {\n",
    "            print(\"running coloc_to_gassocplot via API\")\n",
    "            # query the API if <500 rsids\n",
    "            temp <- coloc_to_gassocplot(out)\n",
    "            # increment counter for number of colocalized eQTLs\n",
    "            colocalized_counter <- colocalized_counter + 1\n",
    "        }\n",
    "        # get the rsids matching the gassocplot plot labels\n",
    "        top_marker_gwas <- extract_top_markers_from_gassocplot_output(temp)[1]\n",
    "        top_marker_eqtl <- extract_top_markers_from_gassocplot_output(temp)[2]\n",
    "\n",
    "        result_raw <- out_PPH4_rawResult[[3]]\n",
    "\n",
    "        top_table <<- get_top_marker_raw_data(top_marker_gwas, top_marker_eqtl, result_raw, PP_H4)\n",
    "        top_table_filename <<- \"eQTLs_colocalized_w_GWAS.txt\"\n",
    "        top_table_path <<- glue::glue(eqtl_outdir, '/', user_outdir, '/', top_table_filename)\n",
    "        # print(top_table_path)\n",
    "        # write the header only once\n",
    "        if (!file.exists(top_table_path)) {\n",
    "            write.table(top_table, top_table_path, sep = \"\\t\", row.names = FALSE, quote = FALSE,col.names = TRUE, append = FALSE)\n",
    "        } else {\n",
    "            write.table(top_table, top_table_path, sep = \"\\t\", row.names = FALSE, quote = FALSE, col.names = FALSE, append = TRUE)\n",
    "        }    \n",
    "        # call customized gassocplot function w/ -log(pval) ceiling removed\n",
    "        plot_associated_signals_and_save(tophit_idx, chromosome, PP_H4, temp, glue::glue(plot_dir,'/', user_outdir))\n",
    "\n",
    "\n",
    "        # increment progress bar\n",
    "        print(glue::glue('working on tophit_idx {tophit_idx} out of {nrow(top)}'))\n",
    "        setProgress(tophit_idx / nrow(top), detail = glue::glue('Processing chr {chromosome}: region {tophit_idx} out of {nrow(top)}'))\n",
    "\n",
    "        } # close loop over each gwas top hit\n",
    "    # redraw the main panel explicitly\n",
    "    imgs <<- list.files(glue::glue(plot_dir,'/', user_outdir), pattern=\".png\", full.names = TRUE)\n",
    "    # str(imgs)\n",
    "    output[[\"slickr\"]] <<- renderSlickR({\n",
    "        slickR(imgs) + settings(slidesToShow = 1, slidesToScroll = 1, lazyLoad = 'anticipated', dots = TRUE, arrows = TRUE, infinite = FALSE)\n",
    "    })\n",
    "#     }) # close withProgress\n",
    "} # close loop over chromosomes\n",
    "\n",
    "clean_up(con) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output resulting plots\n",
    "The image files will be stored in the plots.zip directory. In the app, this file is provided through a download button. For the purposes of this tutorial, one of the example plots in the directory will be printed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if colocalized_counter is 0, print message\n",
    "if (colocalized_counter == 0) {\n",
    "    print(\"No colocalized eQTLs found. Try again with different chromosomes or GWAS datasets.\")\n",
    "    \n",
    "} else {\n",
    "    print(glue::glue(\"Number of colocalized eQTLs: {colocalized_counter}\"))\n",
    "}\n",
    "\n",
    "plot_names <- basename(imgs)\n",
    "# zip plots and store in user directory\n",
    "zip::zip(zipfile = 'plots.zip', files = plot_names, root = glue::glue(plot_dir,'/', user_outdir), recurse = FALSE)\n",
    "zip_path <<- glue::glue(plot_dir,'/', user_outdir, '/plots.zip')\n",
    "        \n",
    "        \n",
    "#Show in notebook one of the gassocplots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# From previous coloc_pipeline.ipynb (to be deleted if unneeded once current algorithm is outlined.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# dir_eqtl <- \"eqtls\"\n",
    "dir_eqtlmaf <- \"eqtl_MAF\"\n",
    "\n",
    "# open parquet files\n",
    "# ds_eQTL <- arrow::open_dataset(dir_eqtl, partitioning = \"SNPChr\")\n",
    "ds_eqtlMAF <- arrow::open_dataset(dir_eqtlmaf, partitioning = \"hg19_chr\")\n",
    "\n",
    "con <- dbConnect(duckdb::duckdb())\n",
    "\n",
    "# register the dataset as a DuckDB table, and give it a name\n",
    "# duckdb::duckdb_register_arrow(con, \"eqtlTable\", ds_eQTL)\n",
    "duckdb::duckdb_register_arrow(con, \"mafTable\", ds_eqtlMAF)\n",
    "\n",
    "# read in the file\n",
    "# query the table for each chromosome separately\n",
    "for (chr in 1:22) {\n",
    "    \n",
    "    # Left join result and 2018-07-18_SNP_AF_for_AlleleB_combined_allele_counts_and_MAF_pos_added.txt  on SNPChr and SNPPos\n",
    "    query <- glue::glue(\n",
    "        \"\n",
    "        SELECT SNP, hg19_pos\n",
    "        FROM 'eqtl_MAF/hg19_chr={chr}/*.parquet'\n",
    "        \"\n",
    "    )\n",
    "    tic()\n",
    "    result <- dbGetQuery(con, query)\n",
    "    toc()\n",
    "    # and print the first 10 rows\n",
    "    print(head(result, 10))\n",
    "}\n",
    "# clean up\n",
    "duckdb_unregister(con, \"eqtlTable\")\n",
    "duckdb_unregister(con, \"mafTable\")\n",
    "dbDisconnect(con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wbc count related to autoimmune disease\n",
    "gwas_dataset <- \"ieu-b-30\"\n",
    "dummy_dataset <- \"ieu-a-7\"\n",
    "gwasinfo(id = as.character(gwas_dataset))\n",
    "# gwasinfo(id=as.character(dummy_dataset))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# queries API from https://gwas.mrcieu.ac.uk/ to:\n",
    "    # 1) grab VCF file and \n",
    "    # 2) identify top GWAS hits\n",
    "    # 3) sort by p-value order (lowest to highest)\n",
    "top <- ieugwasr::tophits(\"ieu-b-30\") %>% arrange(p)\n",
    "print(paste0(dim(top)[1], \" GWAS hits identified\"))\n",
    "head(top)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "window_size <- 50000\n",
    "if (!dir.exists(\"gwas_hits\")) {\n",
    "    dir.create(\"gwas_hits\")\n",
    "}\n",
    "\n",
    "# parallel loop over the top GWAS hits in p-value order and iterate over rows\n",
    "foreach(i = 1:nrow(top)) %dopar% {\n",
    "    # str(top[i,])\n",
    "    # add leading 0s to pval_rank_gwashit so that it always has 3 digits\n",
    "    pval_rank_gwashit <- sprintf(\"%03d\", i)\n",
    "    chr_gwashit <- top[i, ]$chr\n",
    "    pos_gwashit <- top[i, ]$position\n",
    "\n",
    "    # print(paste0(\"Processing GWAS hit \", pval_rank_gwashit, \" of \", nrow(top)))\n",
    "\n",
    "    lower <- pos_gwashit - window_size\n",
    "    upper <- pos_gwashit + window_size\n",
    "    # print(paste(pval_rank_gwashit,\" \",chr_gwashit,\" \",lower,\" \",upper))\n",
    "    chrpos <- paste0(chr_gwashit, \":\", lower, \"-\", upper)\n",
    "    # print(chrpos)\n",
    "\n",
    "    # extract SNPS in the region of top hit\n",
    "    out <- ieugwasr_to_coloc(\n",
    "        id1 = as.character(gwas_dataset),\n",
    "        id2 = as.character(dummy_dataset),\n",
    "        chrompos = as.character(chrpos),\n",
    "        type1 = \"quant\",\n",
    "        # type2 = \"cc\"\n",
    "    )\n",
    "\n",
    "    # get rid of dummy dataset\n",
    "    out <- out[1]$dataset1\n",
    "    # str(out)\n",
    "\n",
    "    # export to file (tab-delimited) in directory named gwas_hits in the following format:\n",
    "    # <pval rank>_<chr>_<lower bound>_<upper bound>.tsv\n",
    "    write.table(out,\n",
    "        file = paste0(\n",
    "            \"gwas_hits/\", pval_rank_gwashit, \"_\", chr_gwashit, \"_\", lower, \"_\", upper, \"_gwas.tsv\"\n",
    "        ),\n",
    "        sep = \"\\t\", quote = FALSE, row.names = FALSE, col.names = TRUE\n",
    "    )\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build eQTL DB\n",
    "# python3 eQTL_build_db.py\n",
    "\n",
    "# Query eQTL DB; Change integer in \"-P0\" to specify the number of threads\n",
    "# find gwas_hits/ -type f -name \"*\" | \\\n",
    "# xargs -P0 -I{} python3 eQTL_query_db.py --input_file {} --db_name eQTLs.db --output_dir eqtl_hits\n",
    "    # can check gnomad if maf assumptions are reasonable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gather up all the gwas and eqtl files\n",
    "gwas_file_list <- list.files(path = \"gwas_hits\", pattern = \".tsv\", full.names = TRUE)\n",
    "eqtl_file_list <- list.files(path = \"output_full\", pattern = \".tsv\", full.names = TRUE)\n",
    "\n",
    "# str(gwas_file_list)\n",
    "# str(eqtl_file_list)\n",
    "\n",
    "# make a directory named output_coloc if it doesn't exist\n",
    "dir.exists(\"output_coloc\") || dir.create(\"output_coloc\")\n",
    "\n",
    "for (i in seq_along(gwas_file_list)) {\n",
    "\n",
    "    # loop over elements in gwas_file_list and eqtl_file_list at the same time\n",
    "    gwas_file <- gwas_file_list[i]\n",
    "    eqtl_file <- eqtl_file_list[i]\n",
    "\n",
    "    gwas_data <- as.list(read.table(file = gwas_file, header = TRUE, sep = \"\\t\"))\n",
    "\n",
    "    # working around formatting issues\n",
    "    eqtl_data <- read.table(file = eqtl_file, header = FALSE, sep = \"\\t\", fill = TRUE)\n",
    "    # drop the last column in eqtl_data - not sure why there is an empty column read in from the file???\n",
    "    eqtl_data <- eqtl_data[, -ncol(eqtl_data)]\n",
    "    # rename the columns by first element\n",
    "    colnames(eqtl_data) <- eqtl_data[1, ]\n",
    "    # drop the first row\n",
    "    eqtl_data <- eqtl_data[-1, ]\n",
    "    # set column data types manually\n",
    "    eqtl_data$pvalues <- as.numeric(eqtl_data$pvalues)\n",
    "    eqtl_data$N <- as.integer(eqtl_data$N)\n",
    "    eqtl_data$MAF <- as.numeric(eqtl_data$MAF)\n",
    "    eqtl_data$beta <- as.numeric(eqtl_data$beta)\n",
    "    eqtl_data$varbeta <- as.numeric(eqtl_data$varbeta)\n",
    "    eqtl_data$type <- as.character(eqtl_data$type)\n",
    "    eqtl_data$snp <- as.character(eqtl_data$snp)\n",
    "    eqtl_data$z <- as.numeric(eqtl_data$z)\n",
    "    eqtl_data$chr <- as.integer(eqtl_data$chr)\n",
    "    eqtl_data$pos <- as.integer(eqtl_data$pos)\n",
    "    eqtl_data$id <- as.character(eqtl_data$id)\n",
    "\n",
    "    eqtl_data <- as.list(eqtl_data)\n",
    "\n",
    "    # run coloc\n",
    "    res <- coloc::coloc.abf(gwas_data, eqtl_data)\n",
    "\n",
    "    #  Get locus from filenames\n",
    "    SNP_locus_name <- basename(gwas_file)\n",
    "    SNP_locus_name <- sub(\"_gwas\\\\..*$\", \"\", SNP_locus_name)\n",
    "    # split on _ assign first element to pval_rank, second to chr, third to lower bound, fourth to upper bound\n",
    "    SNP_locus_name <- strsplit(SNP_locus_name, \"_\")\n",
    "    # flatten to vector\n",
    "    SNP_locus_name <- unlist(SNP_locus_name)\n",
    "    pval_rank <- SNP_locus_name[1]\n",
    "    chr <- SNP_locus_name[2]\n",
    "    lower <- SNP_locus_name[3]\n",
    "    upper <- SNP_locus_name[4]\n",
    "\n",
    "\n",
    "    # write posterior probability output to separate log files\n",
    "    # if H4 >= 80%, send to one file\n",
    "    if (as.numeric(res$summary[\"PP.H4.abf\"]) >= 0.8) {\n",
    "        write.table(paste(pval_rank, chr, lower, upper, as.numeric(res$summary[\"PP.H1.abf\"]), as.numeric(res$summary[\"PP.H2.abf\"]), as.numeric(res$summary[\"PP.H3.abf\"]), as.numeric(res$summary[\"PP.H4.abf\"])),\n",
    "            file = paste0(\n",
    "                \"output_coloc/H4gt80.tsv\"\n",
    "            ),\n",
    "            sep = \"\\t\", quote = FALSE, row.names = FALSE, col.names = FALSE, append = TRUE\n",
    "        )\n",
    "        # if H4 < 80%, send to another file\n",
    "    } else if (as.numeric(res$summary[\"PP.H4.abf\"]) < 0.8) {\n",
    "        write.table(paste(pval_rank, chr, lower, upper, as.numeric(res$summary[\"PP.H1.abf\"]), as.numeric(res$summary[\"PP.H2.abf\"]), as.numeric(res$summary[\"PP.H3.abf\"]), as.numeric(res$summary[\"PP.H4.abf\"])),\n",
    "            file = paste0(\n",
    "                \"output_coloc/H4lt80.tsv\"\n",
    "            ),\n",
    "            sep = \"\\t\", quote = FALSE, row.names = FALSE, col.names = FALSE, append = TRUE\n",
    "        )\n",
    "    }\n",
    "    # write.table(paste(pval_rank,chr,lower,upper,as.numeric(res$summary[\"PP.H1.abf\"]),as.numeric(res$summary[\"PP.H2.abf\"]),as.numeric(res$summary[\"PP.H3.abf\"]),as.numeric(res$summary[\"PP.H4.abf\"])), file = paste0(\"output_coloc/\", \"posteriors.tsv\"), sep = \"\\t\", quote=FALSE, row.names = FALSE, col.names = FALSE, append = TRUE)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the highest priors from the H4 < 80% list to a dataframe named posteriors\n",
    "posteriors <- read.table(file = \"output_coloc/H4lt80.tsv\", header = FALSE, sep = \" \")\n",
    "\n",
    "# manually name columns\n",
    "colnames(posteriors) <- c(\"pval_rank\", \"chr\", \"lower\", \"upper\", \"PP.H1.abf\", \"PP.H2.abf\", \"PP.H3.abf\", \"PP.H4.abf\")\n",
    "\n",
    "# sort posteriors by the last column in descending order\n",
    "posteriors <- posteriors[order(-posteriors$PP.H4.abf), ]\n",
    "\n",
    "# only consider top 10 H4 posterior probabilities and identify gwas and eqtl rows they came from\n",
    "top10H4 <- posteriors[1:10, ]\n",
    "head(top10H4, n = 10L)\n",
    "\n",
    "\n",
    "# manually set the gwas and eqtl file names by concatenating the values of the first 4 columns in top5H4\n",
    "gwas_files <- paste0(\"gwas_hits/\", top10H4$pval_rank, \"_\", top10H4$chr, \"_\", top10H4$lower, \"_\", top10H4$upper, \"_gwas.tsv\")\n",
    "eqtl_files <- paste0(\"eqtl_hits/\", top10H4$pval_rank, \"_\", top10H4$chr, \"_\", top10H4$lower, \"_\", top10H4$upper, \"_eQTLs.tsv\")\n",
    "\n",
    "# loop over records and construct out object for plotting\n",
    "for (i in seq_along(gwas_files)) {\n",
    "\n",
    "    # loop over elements in gwas_file_list and eqtl_file_list at the same time\n",
    "    gwas_file <- gwas_files[i]\n",
    "    eqtl_file <- eqtl_files[i]\n",
    "\n",
    "    gwas_data <- read.table(file = gwas_file, header = TRUE, sep = \"\\t\")\n",
    "\n",
    "    # working around formatting issues\n",
    "    eqtl_data <- read.table(file = eqtl_file, header = FALSE, sep = \"\\t\", fill = TRUE)\n",
    "    # drop the last column in eqtl_data - not sure why there is an empty column read in from the file???\n",
    "    eqtl_data <- eqtl_data[, -ncol(eqtl_data)]\n",
    "    # rename the columns by first element\n",
    "    colnames(eqtl_data) <- eqtl_data[1, ]\n",
    "    # drop the first row\n",
    "    eqtl_data <- eqtl_data[-1, ]\n",
    "    # set column data types manually\n",
    "    eqtl_data$pvalues <- as.numeric(eqtl_data$pvalues)\n",
    "    eqtl_data$N <- as.integer(eqtl_data$N)\n",
    "    eqtl_data$MAF <- as.numeric(eqtl_data$MAF)\n",
    "    eqtl_data$beta <- as.numeric(eqtl_data$beta)\n",
    "    eqtl_data$varbeta <- as.numeric(eqtl_data$varbeta)\n",
    "    eqtl_data$type <- as.character(eqtl_data$type)\n",
    "    eqtl_data$snp <- as.character(eqtl_data$snp)\n",
    "    eqtl_data$z <- as.numeric(eqtl_data$z)\n",
    "    eqtl_data$chr <- as.integer(eqtl_data$chr)\n",
    "    eqtl_data$pos <- as.integer(eqtl_data$pos)\n",
    "    eqtl_data$id <- as.character(eqtl_data$id)\n",
    "\n",
    "    # subset gwas_data and eqtl_data for shared values of chr and pos, drop rows that don't exist in BOTH\n",
    "    gwas_data <- gwas_data[gwas_data$chr %in% eqtl_data$chr & gwas_data$pos %in% eqtl_data$pos, ]\n",
    "    eqtl_data <- eqtl_data[eqtl_data$chr %in% gwas_data$chr & eqtl_data$pos %in% gwas_data$pos, ]\n",
    "\n",
    "    # create a list named out with two elements named dataset1 and dataset 2 and append gwas_data and eqtl_data\n",
    "    out <- list(dataset1 = as.list(gwas_data), dataset2 = as.list(eqtl_data))\n",
    "\n",
    "    # shorten the sublist named id to just 1 element\n",
    "    out$dataset1$id <- out$dataset1$id[1]\n",
    "    out$dataset2$id <- out$dataset2$id[1]\n",
    "    # str(out)\n",
    "\n",
    "    # plot the data\n",
    "    temp <- coloc_to_gassocplot(out)\n",
    "    # str(temp)\n",
    "    gassocplot::stack_assoc_plot(temp$markers, temp$z, temp$corr, traits = temp$traits)\n",
    "}\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "55dd344939daf21fc6a3e99ac35564a930b49a1d935b7f23a278e666455e30f4"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
